File: kotlin/day01/src/main/kotlin/Main.kt
package org.example

import java.io.File
import kotlin.math.abs

fun main() {
    // Part 1.
    val lines = File("input.txt").readLines()
    val lists = lines
        .map { it.split("   ") }
        .map { it[0].toInt() to it[1].toInt() }
        .unzip()
    val (first, second) = lists.first.sorted() to lists.second.sorted()
    val result = first
        .zip(second)
        .sumOf { abs(it.first - it.second) }
    println(result)

    // Part 2.
    val counts = mutableMapOf<Int, Int>()
    for (number in lists.second) {
        counts[number] = counts.getOrDefault(number, 0) + 1
    }
    println(lists.first.sumOf { it * counts.getOrDefault(it, 0) })
}

File: rust/day01/src/main.rs
use std::collections::HashMap;

fn main() {
    let input = std::fs::read_to_string("input.txt").unwrap();
    let (mut left, mut right): (Vec<_>, Vec<_>) = input
        .lines()
        .map(|line| {
            line.split("   ")
                .map(|x| x.parse::<i32>().unwrap())
                .collect::<Vec<_>>()
        })
        .map(|numbers| (numbers[0], numbers[1]))
        .unzip();

    // Part 1.
    left.sort();
    right.sort();
    let result: i32 = left
        .iter()
        .zip(right.iter())
        .map(|(x, y)| (x - y).abs())
        .sum();
    println!("Result: {}", result);

    // Part 2.
    let mut counts = HashMap::new();
    for number in right {
        *counts.entry(number).or_insert(0) += 1;
    }
    let result: i32 = left
        .iter()
        .map(|number| number * counts.get(&number).unwrap_or(&0))
        .sum();
    println!("Result: {}", result);
}
